### PR TITLE
Add contextual overloading

### PR DESCRIPTION
**Pull request by CodaFi**
### What's in this pull request?

Refactor ScopeCheck to not drop namespaces and accept overloads for constructors.  These overloads are required to be disambiguated by contextual types at elaboration time.  We will not be introducing disjunctions into the constraint system!

Haven't touched records, but they will go through the same treatment in a future patch.

### CODE REVIEW

File: Sources/Mantle/Check.swift [hunk 1]
```diff
@@ -455,7 +455,12 @@ extension TypeChecker where PhaseState == CheckPhaseState {
           fatalError()
         }
         return (.absurd, type)
-      case let .constructor(dataCon, synPats):
+      case let .constructor(dataConSet, synPats):
+        guard
+          let dataCon = self.disambiguateConstructor(dataConSet, patType)
+        else {
+          fatalError()
```
Comment (harlanhaskins) [inline Sources/Mantle/Check.swift:462]: Message, please.

File: Sources/Mantle/Elaborate.swift [hunk 2]
```diff
@@ -138,8 +138,19 @@ extension TypeChecker where PhaseState == ElaboratePhaseState {
...
                          TT.refl, from: syntax)
 
-    case let .constructor(dataCon, constructorArgs):
-      let (openedCon, dc) = self.getOpenedDefinition(dataCon)
+    case let .constructor(user, dataConSet, constructorArgs):
+      guard let constr = self.disambiguateConstructor(dataConSet, exType) else {
+        self.engine.diagnose(.useOfAmbiguousConstructor(user.name),
+                             node: user.name.syntax) {
+          for cand in dataConSet {
+            $0.note(.ambiguousConstructorCandidate(cand),
```
Comment (harlanhaskins) [inline Sources/Mantle/Elaborate.swift:146]: 👍👍👍

**Review by harlanhaskins — APPROVED** [review]
LGTM

File: Sources/Mantle/Check.swift (hunk 1 above)
Comment (CodaFi) [reply Sources/Mantle/Check.swift:462]: Can't without refactoring.  This routine "can't fail" given its current type.

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**